### ISSUE TITLE
Upgrading to 1.2.0 is broken on Raspberry PI

### ISSUE THREAD

**Issue by tw-maelstrom** [aws/aws-iot-device-sdk-python#67]
I had the AWS IoT Python SDK v. `1.1.2` installed and running on my Raspberry PI. I upgraded to `1.2.0` and am now getting threading errors when I try to connect. 

Code:
```
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient

myMQTTClient = AWSIoTMQTTClient("Device")
myMQTTClient.configureEndpoint("myendpoint", 8883)
myMQTTClient.configureCredentials("aws-iot-rootCA.crt", "private.pem.key", "certificate.pem.crt")
myMQTTClient.configureOfflinePublishQueueing(-1)  # Infinite offline Publish queueing
myMQTTClient.configureDrainingFrequency(2)  # Draining: 2 Hz
myMQTTClient.configureConnectDisconnectTimeout(10)  # 10 sec
myMQTTClient.configureMQTTOperationTimeout(5)  # 5 sec

myMQTTClient.connect()
```
Errors:
```
Exception in thread Thread-1 (most likely raised during interpreter shutdown):
Exception in thread Thread-3 (most likely raised during interpreter shutdown):
```

**Comment by tw-maelstrom** [comment]
I have played around with this more and am getting more detailed trace back errors, however they are somewhat random.

Here's an example:

```
Exception in thread Thread-1 (most likely raised during interpreter shutdown):Exception in thread Thread-3 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
Traceback (most recent call last):

  File "/usr/lib/python2.7/threading.py", line 1071, in run  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 614, in wait

  File "/usr/lib/python2.7/threading.py", line 355, in wait  File "/usr/lib/python2.7/threading.py", line 754, in run
<type 'exceptions.TypeError'>: 'NoneType' object is not callable

  File "/usr/local/lib/python2.7/dist-packages/AWSIoTPythonSDK/core/protocol/internal/workers.py", line 137, in _dispatch
  File "/usr/lib/python2.7/threading.py", line 289, in __exit__
  File "/usr/lib/python2.7/threading.py", line 216, in __exit__
  File "/usr/lib/python2.7/threading.py", line 203, in release
<type 'exceptions.TypeError'>: 'NoneType' object is not callable
```

I also tried sending a message and sometimes the message was sent even though the threading error happened.

**Comment by pickfire** [comment]
I also got this issue after upgrading to 1.2.0, I think this isn't limited to raspberry pi. The issue occurs at https://github.com/aws/aws-iot-device-sdk-python/blob/master/AWSIoTPythonSDK/core/protocol/internal/workers.py#L225 where the `client` isn't passed to the called function. Maybe because of https://github.com/aws/aws-iot-device-sdk-python/blob/master/AWSIoTPythonSDK/core/protocol/internal/workers.py#L254 ?

**Comment by vareddy** [comment]
Hi @tw-maelstrom,
Apologies for the late reply. Thank you for your interest in AWS IoT!
We were able to reproduce this issue if you have just a connect call and nothing else in your script, when the interpreter shuts down and the threads get killed. Could you add a sleep loop to the end of your program and let us know if the behavior changes?
Please let us know if you have any more questions.
Thanks!
Varun

**Comment by vareddy** [comment]
Hi @pickfire,
We believe your issue might be different from the one seen above. Could you open a new issue detailing your findings?
Thank you for your interest in AWS IoT!
Varun

**Comment by khirstinova** [comment]
I get this during general shutdown even if I have performed other operations and done a disconnect() call:

Exception in thread Thread-1 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 551, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 504, in run
  File "/usr/lib/python2.7/site-packages/AWSIoTPythonSDK/core/protocol/internal/workers.py", line 137, in _dispatch
  File "/usr/lib/python2.7/threading.py", line 214, in __exit__
  File "/usr/lib/python2.7/threading.py", line 154, in __exit__
  File "/usr/lib/python2.7/threading.py", line 141, in release
<type 'exceptions.TypeError'>: 'NoneType' object is not callable

**Comment by morrissinger** [comment]
I had initially opened https://github.com/aws/aws-iot-device-sdk-python/issues/72, but it seems to be a duplicate of this one. If others feel differently, I can reopen #72.

**Comment by toshitanian** [comment]
I got same kind of error.
```
"MqttCore initialized"
"Client id: my-first-device"
"Protocol version: MQTTv3.1.1"
"Authentication type: TLSv1.2 certificate based Mutual Auth."
"Configuring offline requests queueing: max queue size: 0"
"Configuring offline requests queue draining interval: 0.100000 sec"
"Configuring endpoint..."
"Configuring certificates..."
"Configuring reconnect back off timing..."
"Base quiet time: 1.000000 sec"
"Max quiet time: 32.000000 sec"
"Stable connection time: 20.000000 sec"
"Configuring connect/disconnect time out: 10.000000 sec"
"Configuring MQTT operation time out: 5.000000 sec"
"Performing sync connect..."
"Performing async connect..."
"Keep-alive: 30.000000 sec"
"Performing sync subscribe..."
"Subscribed to delta topic for deviceShadow: my-first-device"
"Performing sync subscribe..."
"Performing sync subscribe..."
"Subscribed to get accepted/rejected topics for deviceShadow: my-first-device"
"Performing sync publish..."
Traceback (most recent call last):
  File "/Users/kawasakitoshiya/.pyenv/versions/3.6.0/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/kawasakitoshiya/.pyenv/versions/3.6.0/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
...
<our code>
...
    self.logger('starging agent')
TypeError: 'Logger' object is not callable
Unhandled exception in thread started by <bound method Thread._bootstrap of <Thread(Thread-5, started daemon 123145328578560)>>
```

**Comment by CheetahGod** [comment]
I am having the same problem how do I fix this?

**Comment by liuszeng** [comment]
Hi everyone,

Sorry for the delayed response. I am having some personal issues over the past months and was not available.

The error seems to be related to the client shutdown at the end of the script, where the python runtime starts cleaning up resources upon script exits. In v1.2.0, a disconnect call/exiting the script will not wait until the internal worker thread to stop and therefore results in freeing up resources that are still in use in the running background thread.

One thing you can do now is, upon exiting the script (if you are not doing an infinite loop run), wait for a bit after you called disconnect API call for it to terminate the internal worker thread so that resources are fully shutdown before python runtime starts cleaning them up.

I've created an internal tracker for this and will prioritize it against other tasks. Please let us know you find anything else related to this.

Thanks,
Liusu

**Comment by liuszeng** [comment]
Addressed in v1.3.1.
